### ISSUE TITLE
Support bypass of hostname verification

### ISSUE THREAD

**Issue by mtjhax** [nahi/httpclient#371]
Currently, it is possible to completely disable SSL peer verification by configuring the verify mode as OpenSSL::SSL::VERIFY_NONE.

It would be useful to be able to verify the peer SSL certificate, but skip the post-connection hostname (CN / SAN) checks. This would support use cases where a server's hostname was changed (or in my case, the server is only accessible via IP address) but for some reason updated certificates were not available. The client's connection would be encrypted although obviously less secure, which is preferable to disabling verification completely.

The proposed solution is to implement a new boolean configuration, `HTTPClient::SSLConfig#ssl_verifyhost`, with a default value of true. Setting this value to false would cause `HTTPClient::SSLSocket#post_connection_check` to be skipped.

A possible enhancement might be to allow `ssl_verifyhost` to contain a string, which would be an alternative hostname that is expected to be found in the peer certificate.

I have patched this to work in my code and will submit a merge request pending some additional testing.

**Comment by jrochkind** [comment]
I'm not sure trusting a cert that is signed by a trusted authority for a _different_ hostname than the one you are connecting to... actually gives you any more security in any way than just VERIFY_NONE.  It would mean that any cert at all would work, so long as it was signed by a trusted authority (for some other domain). 

I'd fix your cert, or just use VERIFY_NONE. 

(I am just an interested person, not involved in httpclient development directly in any way)

**Comment by mtjhax** [comment]
@jrochkind I agree that most of the time, not verifying hostname is essentially the same as VERIFY_NONE.

I neglected to mention that in my case, the cert is self-signed and the root CA is pinned on the client side. In theory, you won't be vulnerable to MITM unless the attacker stole the CA cert. Verifying the signature is still useful with cert pinning but the hostname is less important (if an attacker stole the CA, they can make the hostname whatever they want).

I admit this is a somewhat narrow set of use cases to justify adding a feature, but in my case this pops up several times a year when doing trial SaaS integrations through a customer's firewall to a sandbox server on their intranet.
